File: 1678.goal-parser-interpretation.java
/*
 * @lc app=leetcode id=1678 lang=java
 *
 * [1678] Goal Parser Interpretation
 */

// @lc code=start
class Solution {
    public String interpret(String command) {
        char[] buf = command.toCharArray();
        StringBuilder sb = new StringBuilder();
        int idx = 0;
        while (idx != buf.length) {
            if (buf[idx] == 'G') {
                sb.append("G");
                idx += 1;
            } else if (buf[idx] == '(') {
                if (buf[idx + 1] == ')') {
                    sb.append("o");
                    idx += 2;
                } else if (buf[idx + 1] == 'a') {
                    sb.append("al");
                    idx += 4;
                }
            }
        }
        return sb.toString();
    }
}
// @lc code=end

File: 1678.goal-parser-interpretation.cpp
/*
 * @lc app=leetcode id=1678 lang=cpp
 *
 * [1678] Goal Parser Interpretation
 */

// @lc code=start
#include <string>

using namespace std;

class Solution {
public:
    string interpret(string& command) {
        auto r = command.begin(), w = command.begin();
        while (r != command.end()) {
            if (*r == 'G') {
                *(w++) = *r++;
            } else if (*r == '(' && *(r + 1) == ')') {
                *(w++) = 'o'; r += 2;
            } else {
                *(w++) = 'a'; *(w++) = 'l';
                r += 4;
            }
        }
        command.erase(w, command.end());
        return command;
    }
};
// @lc code=end
